pkg/ring/basic_lifecycler.go

```diff
@@ -0,0 +1,474 @@
+			return nil
+
+		case <-heartbeatTicker.C:
+			l.heartbeat(ctx)
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// Verifies that tokens that this instance has registered to the ring still belong to it.
+// Gossiping ring may change the ownership of tokens in case of conflicts.
+// If instance doesn't own its tokens anymore, this method generates new tokens and stores them to the ring.
+func (l *BasicLifecycler) verifyTokens(ctx context.Context) bool {
+	result := false
+
+	err := l.updateInstance(ctx, func(r Desc, i *IngesterDesc) bool {
+		// At this point, we should have the same tokens as we have registered before.
+		actualTokens, takenTokens := r.TokensFor(l.instanceID)
+
+		if actualTokens.Equals(l.GetTokens()) {
+			// Tokens have been verified. No need to change them.
+			result = true
+			return false
+		}
+
+		// uh, oh... our tokens are not our anymore. Let's try new ones.
+		needTokens := l.cfg.NumTokens - len(actualTokens)
+
+		level.Info(l.logger).Log("msg", "generating new tokens", "count", needTokens, "ring", l.ringName)
+		newTokens := GenerateTokens(needTokens, takenTokens)
+
+		actualTokens = append(actualTokens, newTokens...)
+		sort.Sort(actualTokens)
+
+		i.Tokens = actualTokens
+		return true
+	})
+
+	if err != nil {
+		level.Error(l.logger).Log("msg", "failed to verify tokens", "ring", l.ringName, "err", err)
+		return false
+	}
+
+	return result
+}
+
+// unregister removes our entry from the store.
+func (l *BasicLifecycler) unregisterInstance(ctx context.Context) error {
+	level.Info(l.logger).Log("msg", "unregistering instance from ring", "ring", l.ringName)
+
+	err := l.store.CAS(ctx, l.ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		if in == nil {
+			return nil, false, fmt.Errorf("found empty ring when trying to unregister")
+		}
+
+		ringDesc := in.(*Desc)
+		ringDesc.RemoveIngester(l.instanceID)
+		return ringDesc, true, nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	l.currState.Lock()
+	l.currInstanceDesc = nil
+	l.currState.Unlock()
+
+	l.metrics.tokensToOwn.Set(0)
+	l.metrics.tokensOwned.Set(0)
+	return nil
+}
+
+func (l *BasicLifecycler) updateInstance(ctx context.Context, update func(Desc, *IngesterDesc) bool) error {
+	var instanceDesc IngesterDesc
+
+	err := l.store.CAS(ctx, l.ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		ringDesc := getOrCreateRingDesc(in)
+
+		var ok bool
+		instanceDesc, ok = ringDesc.Ingesters[l.instanceID]
+
+		// This could happen if the backend store restarted (and content deleted)
+		// or the instance has been forgotten. In this case, we do re-insert it.
+		if !ok {
+			level.Warn(l.logger).Log("msg", "instance missing in the ring, adding it back", "ring", l.ringName)
+			instanceDesc = ringDesc.AddIngester(l.instanceID, l.instanceAddr, l.instanceZone, l.GetTokens(), l.GetState())
+		}
+
+		changed := update(*ringDesc, &instanceDesc)
+		if !changed {
+			return nil, false, nil
+		}
```

Comment:
If `update` didn't do a change, but there was no entry before, we should still proceed and write updated ring to the store.

Reply:
Thanks for spotting it! I fixed it and added a test to cover it.